posts/inside-rust/2020-06-24-lto-improvements.md

```diff
@@ -0,0 +1,89 @@
+
+## What has changed
+
+Changes have been made to both `rustc` and Cargo to control which libraries should include object code and which should include bitcode based on the project's [profile] LTO settings. If the project is not using LTO, then Cargo will instruct `rustc` to not place bitcode in the rlib files, which should reduce the amount of disk space used. This may have a small improvement in performance since `rustc` no longer needs to compress and write out the bitcode.
+
+If the project is using LTO, then Cargo will instruct `rustc` to not place object code in the rlib files, avoiding the expensive code generation step. This should improve the build time when building from scratch, and reduce the amount of disk space used.
+
+Two `rustc` flags are now available to control how the rlib is constructed:
+
+* [`-C linker-plugin-lto`] causes `rustc` to only place bitcode in the `.o` files, and skips code generation. Cargo uses this when the rlib is only intended for use with LTO. This can also be used when doing cross-language LTO.
+* [`-C embed-bitcode=no`] causes `rustc` to avoid placing bitcode in the rlib altogether. Cargo uses this when LTO is not being used, which reduces some disk space usage.
+
+Additionally, the method in which bitcode is embedded in the rlib has changed. Previously, `rustc` would place compressed bitcode as a `.bc.z` file in the rlib archive. Now, the bitcode is placed as an uncompressed section within each `.o` [object file] in the rlib archive. This avoids a small performance hit for compressing the bitcode, and also matches the standard format used by clang.
+
+## Improvements
+
+The following is a summary of improvements observed on a small number of real-world projects of small and medium size. Improvements of a project will depend heavily on the code, optimization settings, operating system, environment, and hardware. These were recorded with the 2020-06-21 nightly release on Linux with parallel job settings between 2 and 32.
+
+The performance wins for debug builds were anywhere from 0% to 4.7% faster. Larger binary crates tended to fair better than smaller library crates.
+
+LTO builds were recorded anywhere from 4% to 20% faster. Thin LTO faired consistently better than fat LTO.
+
+The number of parallel jobs also had a large impact on the amount of improvement. Lower parallel job counts saw substantially more benefit than higher ones. A project built with `-j2` can be 20% faster, whereas the same project at `-j32` would only be 1% faster. Presumably this is because the code-generation phase benefits from higher concurrency, so it was taking a relatively smaller total percentage of time.
```

Comment:
"Lower parallel job counts saw substantially more benefit than higher ones" -- oh man that's pipelining in action I think!